### ISSUE TITLE
Add GPU Feature Support

### ISSUE THREAD

**Issue by mattleibow** [mono/SkiaSharp#138]
This is going to be the root issue where we discuss how and what we put forward to get the GPU / hardware accelerated drawing in place. Our aim is to make sure everything that needs to be is here, but not adding too much more just because we can. 

The idea behind SkiaSharp is to make sure that we wrap the C++ as nicely as we can, but not create a new engine. If we add lots of code, when Google makes a change, we will then have to go and update all that code. SkiaSharp is supposed to be thin. We can always create a set of extension APIs or packages that make everything awesome.

First things first, we are working in the [gpu-support](https://github.com/mono/SkiaSharp/tree/gpu-support) branch for the C#/managed section and the [gpu-support](https://github.com/mono/skia/tree/gpu-support) branch for the native C++ Skia.

You can view the [managed diff](https://github.com/mono/SkiaSharp/compare/gpu-support) and the [native diff](https://github.com/mono/skia/compare/xamarin-mobile-bindings...mono:gpu-support).

If you think we need an API, just leave a comment and we can make sure that we add it. If you created another issue already, just add a link so we can have everything in one place. If your feature is too big for normal support, just create another issue and leave a comment. It doesn't matter what you do, we will make sure that everything gets organized for you!

**Comment by mattleibow** [comment]
So far, we have these types:
 - `GRContext`
    - `static GRContext Create(GRBackend)`
    - `static GRContext Create(GRBackend, IntPtr)`
    - `static GRContext Create(GRBackend, GRGlInterface)`
 - `GRGlInterface`
    - `static GRGlInterface CreateDefaultInterface()`
    - `static GRGlInterface CreateNativeInterface()`
 - `SKCanvas`
    - `static SKSurface Create(GRContext, GRBackendRenderTargetDesc, SKSurfaceProps)`
    - `static SKSurface Create(GRContext, GRBackendTextureDesc, SKSurfaceProps)`
    - `static SKSurface CreateAsRenderTarget(GRContext, GRBackendTextureDesc, SKSurfaceProps)`
    - `static SKSurface Create(GRContext, bool, SKImageInfo, int, SKSurfaceProps)`
 - `GRBackendRenderTargetDesc`
 - `GRBackendTextureDesc`

And these enums:
 - `SKSurfacePropsFlags`
 - `GRSurfaceOrigin`
 - `GRPixelConfig`
 - `GRBackend`
 - `GRBackendTextureDescFlags`

**Comment by mattleibow** [comment]
I just noticed that although Skia can have an OpenGL backend or a DirectX (through ANGLE) backend, it cannot have both. This is sort of a "problem" on Windows desktop as which is best?

There is a way to have "both" and that is to have two sets of dlls, one with OpenGL and the other DirectX. Then, as a developer, you can set either a MSBuild property (or something similar) and select the desired backend, without having to change the code.

We could also split the NuGet into per-platform packages, and just have two packages for Windows.

**Comment by kekekeks** [comment]
@mattleibow Actually you *can* create GrContext with needed opengl interface using GrGLAssembleGLInterface by passing a pointer to glGetProcAddress from corresponding opengl implementation.
See https://github.com/AvaloniaUI/libperspesk/blob/master/src/gl.cpp#L200

**Comment by kekekeks** [comment]
It may be worth to allow user to manually initialize that from C# code (by passing glGetProcAddress pointer)

**Comment by mattleibow** [comment]
@kekekeks Thanks! I did some work and added code so you could do this:

```csharp
var openGLES = ObjCRuntime.Dlfcn.dlopen("/System/Library/Frameworks/OpenGLES.framework/OpenGLES", 0);
var glInterface = GRGlInterface.AssembleGlesInterface((ctx, name) => {
    return ObjCRuntime.Dlfcn.dlsym(openGLES, name);
});
```

**Comment by jazzay** [comment]
Hello. I have tried the latest Windows GL demo app with latest SkiaSharp 1.53.2-gpu2 nuget and vector drawing operations are all messed up. Simple operations like DrawCircle do not work when bound to a GL Backend. I also tried previous gpu1 nuget and it does not work either. The Shader based demos all work fine. Any eta on when this could be fixed?

**Comment by mattleibow** [comment]
@jazzay I am not able to reproduce any issues. All the demos work on both Mac and Windows. Are you able to indicate which demo and send a screenshot?

**Comment by kekekeks** [comment]
That might be related to particular OpenGL drivers. I'd recommend to use ANGLE on windows instead of relying on proper OpenGL support.

**Comment by jazzay** [comment]
Hey @kekekeks! I assume that is something I must configure on my machine? And then SkiaSharp will just pick up that driver configuration?

**Comment by jazzay** [comment]
Hey I installed the latest NVidia GTX 660 TI driver (my video card), ran GL Bench passing with flying colors. Then build/run exact copy of the source from here: https://github.com/mono/SkiaSharp/tree/gpu-support/samples/Skia.WindowsDesktop.GLDemo. This is what I see for first demo:

![image](https://cloud.githubusercontent.com/assets/554363/17997605/434b9d34-6b24-11e6-9c66-0bf68b3502f2.png)

Perlin noise demo runs fine, and fast indicating HW acceleration:

![image](https://cloud.githubusercontent.com/assets/554363/17997618/61b704d4-6b24-11e6-94b0-4fdbd6642c3e.png)

I doubt this is an issue w my driver/hw setup.

**Comment by mattleibow** [comment]
This appears to be working for most cases, so I w\am merging this into master and see what happens :)

**Comment by DanWBR** [comment]
@mattleibow , where can I find a sample showing how to enable gpu acceleration on Android? thanks.

**Comment by mattleibow** [comment]
Check out this issue #148, it had info and a branch with code

**Comment by DanWBR** [comment]
@mattleibow thanks a lot, got it working after commenting this line (Flush() not found): https://github.com/mono/SkiaSharp/blob/skiasharp-views/views/SkiaSharp.Views.Android/SKGLSurfaceViewRenderer.cs#L28

**Comment by mattleibow** [comment]
@DanWBR that one is not yet released to 1.54.x. Once I get the Windows views done, I will update the SkiaSharp NuGet to v1.54.1 and see if I can get a beta out for the views.

Just a note: I reworked all the iOS views/layers (and I think they are better for it)
